p4include/v1model.p4

```diff
@@ -56,23 +56,71 @@ struct standard_metadata_t {
     //
     // flattening fields that exist in bmv2-ss
     // queueing metadata
-    @alias("queueing_metadata.enq_timestamp") bit<32> enq_timestamp;
-    @alias("queueing_metadata.enq_qdepth")    bit<19> enq_qdepth;
-    @alias("queueing_metadata.deq_timedelta") bit<32> deq_timedelta;
-    @alias("queueing_metadata.deq_qdepth")    bit<19> deq_qdepth;
+    @alias("queueing_metadata.enq_timestamp")
+    bit<32> enq_timestamp;
+    @alias("queueing_metadata.enq_qdepth")
+    bit<19> enq_qdepth;
+    @alias("queueing_metadata.deq_timedelta")
+    bit<32> deq_timedelta;
+    /// queue depth at the packet dequeue time.
+    @alias("queueing_metadata.deq_qdepth")
+    bit<19> deq_qdepth;
+
     // intrinsic metadata
```

Comment:
I would recommend not adding standard metadata fields to v1model just because they are used exclusively in switch.p4_14:
 * makes things more complex
 * leads to some non-sensical situations (what's the difference between ucast_egress_port and egress_spec / egress_port)
 * switch.p4 is not maintained anymore, it is in a frozen state and people compiling it are very likely to be compiling it in the context of p4lang/p4factory with the legacy P4_14 compiler
 * I know it was mentioned at the meeting that we use some older copy of switch.p4 for compiler regressions, but at this stage I would imagine we have good enough coverage without it

It is probably a good opportunity to remove v1model standard metadata fields which are used internally by bmv2 simple_switch and should not be programmed directly by the user: clone_spec, lf_field_list, resubmit_flag, recirculate_flag. That would also resolve this issue: https://github.com/p4lang/p4c/issues/1845. The bmv2 backend can synthesize these fields when generating the JSON.

Reply:
I would prefer if we could surgically edit switch.p4 to remove the use of these fields. Then we can keep the remaining program, and hopefully it still offers some value. But I haven't looked to see how they are used.

Reply:
I tried that, but it does not look particularly easy; these fields are used in a bunch of places.
The compiler is already quite slow, but we should have at least one large program to make sure we don't make it unacceptably slow. For example, on a large program a quadratic-time algorithm will show up clearly.

Reply:
Would fabric.p4 (https://github.com/opennetworkinglab/onos/blob/master/pipelines/fabric/src/main/resources/fabric.p4) with all features enables (INT & SPGW) be an acceptable replacement? It is not as big, but at least it is actively maintained.

Reply:
We should have fabric even if we keep switch.p4. 
If we can make it run some STF tests then it may be a good enough replacement.
I won't object to deleting switch.p4.

Reply:
I have just gone through these new fields to see which might be easy to eliminate in the p4c test programs, if people are interested in trying to do that.

I will start by listing the easy ones to get rid of.  Later below I will describe the ones that are only removable with changes to the test programs that I don't know the impact of.  It kind of depends on whether anyone actually uses the 2 versions of switch.p4 in the p4lang/p4c repository for anything other than testing p4c.

+ ucast_egress_port

only occurs in in this one test program: p4_14_samples/sai_p4.p4  Easily replaced with standard_metadata.egress_spec

+ ingress_global_tstamp

Can easily be replaced with intrinsic_metadata.ingress_global_timestamp, which is the current name used in v1model.p4.  Same bit width, too.

+ enq_tstamp

Easily replaceable with queueing_metadata.enq_timestamp, which v1model already has.

+ enq_congest_stat and deq_congest_stat

switch.p4 programs do not mention this field other than to define it.  Thus easily removed.

+ mcast_hash

Defined in about a dozen test programs, P4_14 and P4_16, and both switch.p4 versions.  It is only assigned a value in one or two places in switch.p4, and is an output of the ingress control, intended to be used by the packet buffer / traffic manager.  simple_switch ignores it, so deleting it and commenting out (or perhaps #ifdef-ing out) those few assignments to it would be a very small change to switch.p4.

+ deflection_flag and deflect_on_drop

Only used in the two versions of switch.p4.  There is nothing similar to these in v1model to replace them with.  They could be #ifdef'd out, as suggested for mcast_hash, since they seem to be used in only a few lines of code.

I could make a PR with the changes suggested above, in case people think they would be useful to look at and consider merging.

Reply:
That is very impressive! If indeed these are the only uses then we could potentially keep switch.p4 as well. I looked at the hash, and it seemed to have a bunch of uses, so I got discouraged.